File: mycelix-core/libs/fl-aggregator/src/identity/assurance.rs
//! Assurance Level Calculation for Multi-Factor Identity
//! Aligned with Epistemic Charter v2.0 E-Axis
//!
//! Maps identity factor verification to graduated capabilities

use crate::identity::{AssuranceLevel, FactorCategory, FactorStatus, IdentityFactor};
use std::collections::{HashMap, HashSet};

/// Result of assurance level calculation
#[derive(Debug, Clone)]
pub struct AssuranceResult {
    /// Calculated assurance level
    pub level: AssuranceLevel,

    /// Numerical score (0.0-1.0)
    pub score: f32,

    /// Active factor IDs
    pub active_factors: Vec<String>,

    /// Factor categories present
    pub categories_present: HashSet<FactorCategory>,

    /// Capabilities granted at this level
    pub capabilities: HashSet<Capability>,

    /// Recommendations for improvement
    pub recommendations: Vec<String>,
}

/// Capabilities available at different assurance levels
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Capability {
    // E0 capabilities
    ReadPublicData,
    BrowseAnonymous,

    // E1 capabilities
    PostComments,
    SendMessages,
    CreateContent,

    // E2 capabilities
    ParticipateCommunity,
    SubmitProposals,
    ReceiveRewards,
    ParticipateFL,

    // E3 capabilities
    VoteGovernance,
    ValidateClaims,
    RunValidator,

    // E4 capabilities
    ProposeConstitutionalAmendment,
    JoinKnowledgeCouncil,
    ServeGuardian,
    ManageTreasury,
}

impl Capability {
    /// Get capabilities for each assurance level
    pub fn for_level(level: AssuranceLevel) -> HashSet<Capability> {
        use Capability::*;

        let mut caps = HashSet::new();

        // E0 - Anonymous
        caps.insert(ReadPublicData);
        caps.insert(BrowseAnonymous);

        if level >= AssuranceLevel::E1 {
            caps.insert(PostComments);
            caps.insert(SendMessages);
            caps.insert(CreateContent);
        }

        if level >= AssuranceLevel::E2 {
            caps.insert(ParticipateCommunity);
            caps.insert(SubmitProposals);
            caps.insert(ReceiveRewards);
            caps.insert(ParticipateFL);
        }

        if level >= AssuranceLevel::E3 {
            caps.insert(VoteGovernance);
            caps.insert(ValidateClaims);
            caps.insert(RunValidator);
        }

        if level >= AssuranceLevel::E4 {
            caps.insert(ProposeConstitutionalAmendment);
            caps.insert(JoinKnowledgeCouncil);
            caps.insert(ServeGuardian);
            caps.insert(ManageTreasury);
        }

        caps
    }

    /// Get human-readable description
    pub fn description(&self) -> &'static str {
        use Capability::*;
        match self {
            ReadPublicData => "Read public data and content",
            BrowseAnonymous => "Browse anonymously without account",
            PostComments => "Post comments on discussions",
            SendMessages => "Send direct messages to other users",
            CreateContent => "Create and publish content",
            ParticipateCommunity => "Participate in community activities",
            SubmitProposals => "Submit proposals for consideration",
            ReceiveRewards => "Receive reputation and credit rewards",
            ParticipateFL => "Participate in federated learning rounds",
            VoteGovernance => "Vote on governance proposals",
            ValidateClaims => "Validate epistemic claims",
            RunValidator => "Run a network validator node",
            ProposeConstitutionalAmendment => "Propose constitutional amendments",
            JoinKnowledgeCouncil => "Join the Knowledge Council",
            ServeGuardian => "Serve as a recovery guardian",
            ManageTreasury => "Participate in treasury management",
        }
    }
}

/// Calculate assurance level from active identity factors
///
/// Algorithm:
/// 1. Sum contributions from all ACTIVE factors
/// 2. Apply bonus for factor diversity (multiple categories)
/// 3. Map score to assurance level
/// 4. Return capabilities and recommendations
pub fn calculate_assurance_level(factors: &[&dyn IdentityFactor]) -> AssuranceResult {
    // Filter to active factors
    let active_factors: Vec<_> = factors
        .iter()
        .filter(|f| f.status() == FactorStatus::Active)
        .collect();

    if active_factors.is_empty() {
        return AssuranceResult {
            level: AssuranceLevel::E0,
            score: 0.0,
            active_factors: vec![],
            categories_present: HashSet::new(),
            capabilities: Capability::for_level(AssuranceLevel::E0),
            recommendations: vec![
                "Create a DID with cryptographic key pair".to_string(),
                "Verify your Gitcoin Passport for community participation".to_string(),
                "Set up social recovery guardians".to_string(),
            ],
        };
    }

    // Calculate base score from factor contributions
    let base_score: f32 = active_factors.iter().map(|f| f.contribution()).sum();

    // Get unique categories
    let categories: HashSet<FactorCategory> = active_factors.iter().map(|f| f.category()).collect();

    // Apply diversity bonus: 5% per unique category
    let diversity_bonus = categories.len() as f32 * 0.05;

    // Final score (capped at 1.0)
    let final_score = (base_score + diversity_bonus).min(1.0);

    // Map score to level
    let level = AssuranceLevel::from_score(final_score);

    // Generate recommendations
    let recommendations = generate_recommendations(&active_factors, level);

    AssuranceResult {
        level,
        score: final_score,
        active_factors: active_factors.iter().map(|f| f.factor_id().to_string()).collect(),
        categories_present: categories,
        capabilities: Capability::for_level(level),
        recommendations,
    }
}

/// Generate recommendations for improving assurance level
fn generate_recommendations(
    active_factors: &[&&dyn IdentityFactor],
    current_level: AssuranceLevel,
) -> Vec<String> {
    let mut recommendations = Vec::new();

    // Check what factor types are present
    let factor_types: HashSet<_> = active_factors.iter().map(|f| f.factor_type()).collect();

    // Missing factor recommendations
    if !factor_types.contains("CryptoKey") {
        recommendations.push("CRITICAL: Generate a primary cryptographic key pair".to_string());
    }

    if !factor_types.contains("GitcoinPassport") && current_level < AssuranceLevel::E3 {
        recommendations
            .push("Verify your Gitcoin Passport (required for governance voting)".to_string());
    }

    if !factor_types.contains("SocialRecovery") {
        recommendations
            .push("Set up social recovery with 5+ guardians for account security".to_string());
    }

    if !factor_types.contains("HardwareKey") && current_level >= AssuranceLevel::E2 {
        recommendations.push("Add a hardware security key for enhanced security".to_string());
    }

    if !factor_types.contains("Biometric") && current_level >= AssuranceLevel::E3 {
        recommendations.push("Consider adding biometric verification for convenience".to_string());
    }

    // Level-specific recommendations
    match current_level {
        AssuranceLevel::E0 => {
            recommendations.push("Get started: Create your first DID to join the network".to_string());
        }
        AssuranceLevel::E1 => {
            recommendations
                .push("Reach E2: Add 2+ additional factors for community participation".to_string());
        }
        AssuranceLevel::E2 => {
            recommendations
                .push("Reach E3: Verify Gitcoin Passport >= 20 for governance voting".to_string());
        }
        AssuranceLevel::E3 => {
            recommendations.push(
                "Reach E4: Add hardware key + biometric for constitutional proposals".to_string(),
            );
        }
        AssuranceLevel::E4 => {
            recommendations.push(
                "Maximum verification achieved! You can participate in all governance activities"
                    .to_string(),
            );
        }
    }

    recommendations
}

/// Get the minimum assurance level required for a capability
pub fn get_required_level(capability: Capability) -> AssuranceLevel {
    use Capability::*;

    match capability {
        ReadPublicData | BrowseAnonymous => AssuranceLevel::E0,
        PostComments | SendMessages | CreateContent => AssuranceLevel::E1,
        ParticipateCommunity | SubmitProposals | ReceiveRewards | ParticipateFL => {
            AssuranceLevel::E2
        }
        VoteGovernance | ValidateClaims | RunValidator => AssuranceLevel::E3,
        ProposeConstitutionalAmendment | JoinKnowledgeCouncil | ServeGuardian | ManageTreasury => {
            AssuranceLevel::E4
        }
    }
}

/// Check if an identity has a specific capability
pub fn has_capability(result: &AssuranceResult, capability: Capability) -> bool {
    result.capabilities.contains(&capability)
}

/// Get a summary of capabilities for display
pub fn get_capabilities_summary(result: &AssuranceResult) -> HashMap<String, serde_json::Value> {
    let mut summary = HashMap::new();

    summary.insert("level".to_string(), serde_json::json!(result.level.name()));
    summary.insert("score".to_string(), serde_json::json!(result.score));
    summary.insert(
        "active_factors".to_string(),
        serde_json::json!(result.active_factors.len()),
    );

    let caps = serde_json::json!({
        "read": has_capability(result, Capability::ReadPublicData),
        "write": has_capability(result, Capability::CreateContent),
        "community": has_capability(result, Capability::ParticipateCommunity),
        "fl_participation": has_capability(result, Capability::ParticipateFL),
        "governance": has_capability(result, Capability::VoteGovernance),
        "constitutional": has_capability(result, Capability::ProposeConstitutionalAmendment),
    });

    summary.insert("capabilities".to_string(), caps);

    if let Some(next) = result.level.next() {
        summary.insert("next_level".to_string(), serde_json::json!(next.name()));
    } else {
        summary.insert(
            "next_level".to_string(),
            serde_json::json!("Maximum level achieved"),
        );
    }

    summary.insert(
        "recommendations".to_string(),
        serde_json::json!(result.recommendations),
    );

    summary
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::identity::factors::{CryptoKeyFactor, GitcoinPassportFactor};

    #[test]
    fn test_no_factors_e0() {
        let factors: Vec<&dyn IdentityFactor> = vec![];
        let result = calculate_assurance_level(&factors);

        assert_eq!(result.level, AssuranceLevel::E0);
        assert_eq!(result.score, 0.0);
        assert!(result.active_factors.is_empty());
    }

    #[test]
    fn test_single_crypto_key_e1() {
        let mut crypto = CryptoKeyFactor::generate().unwrap();
        crypto.mark_verified();

        let factors: Vec<&dyn IdentityFactor> = vec![&crypto];
        let result = calculate_assurance_level(&factors);

        // 0.5 contribution + 0.05 diversity = 0.55 -> E2
        assert!(result.level >= AssuranceLevel::E1);
        assert_eq!(result.active_factors.len(), 1);
    }

    #[test]
    fn test_multi_factor_e2() {
        let mut crypto = CryptoKeyFactor::generate().unwrap();
        crypto.mark_verified();

        let mut passport = GitcoinPassportFactor::new("0x123", 25.0).unwrap();
        passport.mark_verified();

        let factors: Vec<&dyn IdentityFactor> = vec![&crypto, &passport];
        let result = calculate_assurance_level(&factors);

        // 0.5 + 0.3 = 0.8 contribution + 0.10 diversity (2 categories) = 0.9 -> E4
        assert!(result.level >= AssuranceLevel::E2);
        assert_eq!(result.categories_present.len(), 2);
    }

    #[test]
    fn test_capabilities_for_level() {
        let e0_caps = Capability::for_level(AssuranceLevel::E0);
        assert!(e0_caps.contains(&Capability::ReadPublicData));
        assert!(!e0_caps.contains(&Capability::VoteGovernance));

        let e3_caps = Capability::for_level(AssuranceLevel::E3);
        assert!(e3_caps.contains(&Capability::VoteGovernance));
        assert!(e3_caps.contains(&Capability::ParticipateFL));
        assert!(!e3_caps.contains(&Capability::ProposeConstitutionalAmendment));

        let e4_caps = Capability::for_level(AssuranceLevel::E4);
        assert!(e4_caps.contains(&Capability::ProposeConstitutionalAmendment));
    }

    #[test]
    fn test_required_level_for_capability() {
        assert_eq!(
            get_required_level(Capability::ReadPublicData),
            AssuranceLevel::E0
        );
        assert_eq!(
            get_required_level(Capability::PostComments),
            AssuranceLevel::E1
        );
        assert_eq!(
            get_required_level(Capability::ParticipateFL),
            AssuranceLevel::E2
        );
        assert_eq!(
            get_required_level(Capability::VoteGovernance),
            AssuranceLevel::E3
        );
        assert_eq!(
            get_required_level(Capability::ProposeConstitutionalAmendment),
            AssuranceLevel::E4
        );
    }

    #[test]
    fn test_diversity_bonus() {
        // Single category: 1 * 0.05 = 0.05 bonus
        let mut crypto = CryptoKeyFactor::generate().unwrap();
        crypto.mark_verified();

        let factors1: Vec<&dyn IdentityFactor> = vec![&crypto];
        let result1 = calculate_assurance_level(&factors1);

        // Two categories: 2 * 0.05 = 0.10 bonus
        let mut passport = GitcoinPassportFactor::new("0x123", 25.0).unwrap();
        passport.mark_verified();

        let factors2: Vec<&dyn IdentityFactor> = vec![&crypto, &passport];
        let result2 = calculate_assurance_level(&factors2);

        // Check diversity bonus effect
        assert!(result2.score > result1.score);
        assert_eq!(result2.categories_present.len(), 2);
    }
}

File: mycelix-workspace/mycelix-core/0TML/src/zerotrustml/identity/assurance.py
# Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
"""
Assurance Level Calculation for Multi-Factor Identity
Aligned with Epistemic Charter v2.0 E-Axis

Maps identity factor verification to graduated capabilities
"""

from enum import Enum
from dataclasses import dataclass
from typing import List, Dict, Set
from .factors import IdentityFactor, FactorStatus


class AssuranceLevel(Enum):
    """
    Assurance levels aligned with Epistemic Charter v2.0 E-Axis

    E0: Anonymous (0.0-0.2) - Unverified, read-only access
    E1: Testimonial (0.2-0.4) - Basic key pair, limited interactions
    E2: Privately Verifiable (0.4-0.6) - Multi-factor, community participation
    E3: Cryptographically Proven (0.6-0.8) - Strong verification, governance voting
    E4: Constitutionally Critical (0.8-1.0) - Maximum verification, constitutional proposals
    """
    E0_ANONYMOUS = "E0_Anonymous"
    E1_TESTIMONIAL = "E1_Testimonial"
    E2_PRIVATELY_VERIFIABLE = "E2_PrivatelyVerifiable"
    E3_CRYPTOGRAPHICALLY_PROVEN = "E3_CryptographicallyProven"
    E4_CONSTITUTIONALLY_CRITICAL = "E4_ConstitutionallyCritical"


@dataclass
class AssuranceResult:
    """
    Result of assurance level calculation

    Contains the calculated level and associated capabilities
    """
    level: AssuranceLevel
    score: float  # 0.0-1.0
    active_factors: List[str]
    capabilities: Set[str]
    recommendations: List[str]


# Capability definitions for each assurance level
LEVEL_CAPABILITIES = {
    AssuranceLevel.E0_ANONYMOUS: {
        "read_public_data",
        "browse_anonymous",
    },

    AssuranceLevel.E1_TESTIMONIAL: {
        "read_public_data",
        "browse_anonymous",
        "post_comments",
        "send_messages",
        "create_content",
    },

    AssuranceLevel.E2_PRIVATELY_VERIFIABLE: {
        "read_public_data",
        "browse_anonymous",
        "post_comments",
        "send_messages",
        "create_content",
        "participate_community",
        "submit_proposals",
        "receive_rewards",
    },

    AssuranceLevel.E3_CRYPTOGRAPHICALLY_PROVEN: {
        "read_public_data",
        "browse_anonymous",
        "post_comments",
        "send_messages",
        "create_content",
        "participate_community",
        "submit_proposals",
        "receive_rewards",
        "vote_governance",
        "validate_claims",
        "run_validator",
    },

    AssuranceLevel.E4_CONSTITUTIONALLY_CRITICAL: {
        "read_public_data",
        "browse_anonymous",
        "post_comments",
        "send_messages",
        "create_content",
        "participate_community",
        "submit_proposals",
        "receive_rewards",
        "vote_governance",
        "validate_claims",
        "run_validator",
        "propose_constitutional_amendment",
        "join_knowledge_council",
        "serve_guardian",
    },
}


def calculate_assurance_level(factors: List[IdentityFactor]) -> AssuranceResult:
    """
    Calculate assurance level from active identity factors

    Algorithm:
    1. Sum contributions from all ACTIVE factors
    2. Apply bonus for factor diversity (multiple categories)
    3. Map score to assurance level
    4. Return capabilities and recommendations

    Args:
        factors: List of identity factors

    Returns:
        AssuranceResult with level, score, capabilities, and recommendations
    """
    # Filter to active factors only
    active_factors = [f for f in factors if f.status == FactorStatus.ACTIVE]

    if not active_factors:
        return AssuranceResult(
            level=AssuranceLevel.E0_ANONYMOUS,
            score=0.0,
            active_factors=[],
            capabilities=LEVEL_CAPABILITIES[AssuranceLevel.E0_ANONYMOUS],
            recommendations=[
                "Create a DID with cryptographic key pair",
                "Verify your Gitcoin Passport for community participation",
                "Set up social recovery guardians"
            ]
        )

    # Calculate base score from factor contributions
    base_score = sum(f.get_contribution() for f in active_factors)

    # Apply diversity bonus
    categories = {f.category for f in active_factors}
    diversity_bonus = len(categories) * 0.05  # 5% per unique category

    # Final score (capped at 1.0)
    final_score = min(base_score + diversity_bonus, 1.0)

    # Map score to assurance level
    level = _score_to_level(final_score)

    # Get capabilities for this level
    capabilities = LEVEL_CAPABILITIES[level]

    # Generate recommendations for improvement
    recommendations = _generate_recommendations(active_factors, level)

    return AssuranceResult(
        level=level,
        score=final_score,
        active_factors=[f.factor_id for f in active_factors],
        capabilities=capabilities,
        recommendations=recommendations
    )


def _score_to_level(score: float) -> AssuranceLevel:
    """
    Map numerical score to assurance level

    Thresholds:
    - 0.0-0.2: E0 (Anonymous)
    - 0.2-0.4: E1 (Testimonial)
    - 0.4-0.6: E2 (Privately Verifiable)
    - 0.6-0.8: E3 (Cryptographically Proven)
    - 0.8-1.0: E4 (Constitutionally Critical)
    """
    if score >= 0.8:
        return AssuranceLevel.E4_CONSTITUTIONALLY_CRITICAL
    elif score >= 0.6:
        return AssuranceLevel.E3_CRYPTOGRAPHICALLY_PROVEN
    elif score >= 0.4:
        return AssuranceLevel.E2_PRIVATELY_VERIFIABLE
    elif score >= 0.2:
        return AssuranceLevel.E1_TESTIMONIAL
    else:
        return AssuranceLevel.E0_ANONYMOUS


def _generate_recommendations(
    active_factors: List[IdentityFactor],
    current_level: AssuranceLevel
) -> List[str]:
    """
    Generate recommendations to improve assurance level

    Args:
        active_factors: Currently active factors
        current_level: Current assurance level

    Returns:
        List of recommended actions
    """
    recommendations = []

    # Check what factor types are present
    factor_types = {f.factor_type for f in active_factors}

    # Recommendations based on missing factors
    if "CryptoKey" not in factor_types:
        recommendations.append("⚠️  CRITICAL: Generate a primary cryptographic key pair")

    if "GitcoinPassport" not in factor_types and current_level.value < AssuranceLevel.E3_CRYPTOGRAPHICALLY_PROVEN.value:
        recommendations.append("📊 Verify your Gitcoin Passport (required for governance voting)")

    if "SocialRecovery" not in factor_types:
        recommendations.append("👥 Set up social recovery with 5+ guardians for account security")

    if "HardwareKey" not in factor_types and current_level.value >= AssuranceLevel.E2_PRIVATELY_VERIFIABLE.value:
        recommendations.append("🔐 Add a hardware security key for enhanced security")

    if "Biometric" not in factor_types and current_level.value >= AssuranceLevel.E3_CRYPTOGRAPHICALLY_PROVEN.value:
        recommendations.append("🔍 Consider adding biometric verification for convenience")

    # Level-specific recommendations
    if current_level == AssuranceLevel.E0_ANONYMOUS:
        recommendations.append("🚀 Get started: Create your first DID to join the network")

    elif current_level == AssuranceLevel.E1_TESTIMONIAL:
        recommendations.append("⬆️  Reach E2: Add 2+ additional factors for community participation")

    elif current_level == AssuranceLevel.E2_PRIVATELY_VERIFIABLE:
        recommendations.append("⬆️  Reach E3: Verify Gitcoin Passport ≥20 for governance voting")

    elif current_level == AssuranceLevel.E3_CRYPTOGRAPHICALLY_PROVEN:
        recommendations.append("⬆️  Reach E4: Add hardware key + biometric for constitutional proposals")

    elif current_level == AssuranceLevel.E4_CONSTITUTIONALLY_CRITICAL:
        recommendations.append("🎉 Maximum verification achieved! You can participate in all governance activities")

    return recommendations


def get_required_level(capability: str) -> AssuranceLevel:
    """
    Get the minimum assurance level required for a capability

    Args:
        capability: Capability name

    Returns:
        Minimum required assurance level
    """
    for level, caps in LEVEL_CAPABILITIES.items():
        if capability in caps:
            return level

    # If not found, require maximum level
    return AssuranceLevel.E4_CONSTITUTIONALLY_CRITICAL


def has_capability(assurance_result: AssuranceResult, capability: str) -> bool:
    """
    Check if an identity has a specific capability

    Args:
        assurance_result: Result from calculate_assurance_level
        capability: Capability to check

    Returns:
        True if identity has this capability
    """
    return capability in assurance_result.capabilities


def get_capabilities_summary(assurance_result: AssuranceResult) -> Dict[str, any]:
    """
    Get a human-readable summary of capabilities

    Args:
        assurance_result: Result from calculate_assurance_level

    Returns:
        Dictionary with capability categories and details
    """
    return {
        "level": assurance_result.level.value,
        "score": round(assurance_result.score, 2),
        "active_factors": len(assurance_result.active_factors),
        "capabilities": {
            "read": has_capability(assurance_result, "read_public_data"),
            "write": has_capability(assurance_result, "create_content"),
            "community": has_capability(assurance_result, "participate_community"),
            "governance": has_capability(assurance_result, "vote_governance"),
            "constitutional": has_capability(assurance_result, "propose_constitutional_amendment"),
        },
        "next_level": _get_next_level(assurance_result.level),
        "recommendations": assurance_result.recommendations
    }


def _get_next_level(current_level: AssuranceLevel) -> str:
    """Get the next assurance level"""
    level_order = [
        AssuranceLevel.E0_ANONYMOUS,
        AssuranceLevel.E1_TESTIMONIAL,
        AssuranceLevel.E2_PRIVATELY_VERIFIABLE,
        AssuranceLevel.E3_CRYPTOGRAPHICALLY_PROVEN,
        AssuranceLevel.E4_CONSTITUTIONALLY_CRITICAL,
    ]

    try:
        current_idx = level_order.index(current_level)
        if current_idx < len(level_order) - 1:
            return level_order[current_idx + 1].value
        return "Maximum level achieved"
    except ValueError:
        return "Unknown"


# Example usage
if __name__ == "__main__":
    from .factors import CryptoKeyFactor, GitcoinPassportFactor, SocialRecoveryFactor
    from .factors import FactorCategory, FactorStatus
    from cryptography.hazmat.primitives.asymmetric import ed25519
    from cryptography.hazmat.primitives import serialization

    # Scenario 1: New user with just crypto key (E1)
    print("=== Scenario 1: New User (Crypto Key Only) ===")
    private_key = ed25519.Ed25519PrivateKey.generate()
    public_key = private_key.public_key()

    crypto_factor = CryptoKeyFactor(
        factor_id="crypto-001",
        factor_type="CryptoKey",
        category=FactorCategory.PRIMARY,
        status=FactorStatus.ACTIVE,
        public_key=public_key.public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw
        )
    )

    result = calculate_assurance_level([crypto_factor])
    print(f"Level: {result.level.value}")
    print(f"Score: {result.score}")
    print(f"Can vote in governance: {has_capability(result, 'vote_governance')}")
    print(f"Recommendations: {result.recommendations[:2]}")

    # Scenario 2: Active community member (E2)
    print("\n=== Scenario 2: Community Member (Crypto + Passport) ===")
    passport_factor = GitcoinPassportFactor(
        factor_id="gitcoin-001",
        factor_type="GitcoinPassport",
        category=FactorCategory.REPUTATION,
        status=FactorStatus.ACTIVE,
        passport_address="0x1234...",
        score=25.0
    )

    result = calculate_assurance_level([crypto_factor, passport_factor])
    print(f"Level: {result.level.value}")
    print(f"Score: {result.score}")
    print(f"Can vote in governance: {has_capability(result, 'vote_governance')}")
    print(f"Can participate in community: {has_capability(result, 'participate_community')}")

    # Scenario 3: Governance participant (E3)
    print("\n=== Scenario 3: Governance Participant (Multi-Factor) ===")
    recovery_factor = SocialRecoveryFactor(
        factor_id="recovery-001",
        factor_type="SocialRecovery",
        category=FactorCategory.SOCIAL,
        status=FactorStatus.ACTIVE,
        threshold=3,
        guardian_dids=["did:mycelix:g1", "did:mycelix:g2", "did:mycelix:g3",
                      "did:mycelix:g4", "did:mycelix:g5"]
    )

    result = calculate_assurance_level([crypto_factor, passport_factor, recovery_factor])
    print(f"Level: {result.level.value}")
    print(f"Score: {result.score}")
    print(f"Can vote in governance: {has_capability(result, 'vote_governance')}")
    print(f"Can propose amendments: {has_capability(result, 'propose_constitutional_amendment')}")

    # Summary
    print("\n=== Capabilities Summary ===")
    summary = get_capabilities_summary(result)
    print(f"Level: {summary['level']}")
    print(f"Score: {summary['score']}")
    print(f"Capabilities: {summary['capabilities']}")
    print(f"Next level: {summary['next_level']}")
